File: src/main/scala/org/xbery/artbeams/users/repository/RoleRepository.scala
package org.xbery.artbeams.users.repository

import java.util.Arrays

import javax.inject.Inject
import javax.sql.DataSource
import org.springframework.stereotype.Repository
import org.xbery.artbeams.common.assets.repository.AssetRepository
import org.xbery.artbeams.users.domain.Role
import org.xbery.overview.{Order, Overview}

/**
  * Role repository.
  * @author Radek Beran
  */
@Repository
class RoleRepository @Inject()(dataSource: DataSource) extends AssetRepository[Role, RoleFilter](dataSource, RoleMapper.Instance) {
  protected lazy val DefaultOrdering = Arrays.asList(new Order(RoleMapper.Instance.nameAttr, false))

  def findRoles(): Seq[Role] = {
    this.findByOverviewAsSeq(new Overview(RoleFilter.Empty, DefaultOrdering))
  }

  def findRolesOfUser(userId: String): Seq[Role] = {
    this.findByOverviewAsSeq(new Overview(RoleFilter.Empty.copy(userId = Some(userId)), DefaultOrdering))
  }

  def updateRolesOfUser(userId: String, roles: Seq[Role]): Unit = {
    this.updateAttributeValues("DELETE FROM user_role WHERE user_id = ?", Arrays.asList[AnyRef](userId))
    for {
      role <- roles
    } {
      this.updateAttributeValues("INSERT INTO user_role (user_id, role_id) VALUES (?, ?)", Arrays.asList[AnyRef](userId, role.id))
    }
  }
}

File: src/main/kotlin/org/xbery/artbeams/users/repository/RoleRepository.kt
package org.xbery.artbeams.users.repository

import org.jooq.DSLContext
import org.jooq.Field
import org.jooq.Table
import org.springframework.stereotype.Repository
import org.xbery.artbeams.common.assets.repository.AssetRepository
import org.xbery.artbeams.jooq.schema.tables.records.RolesRecord
import org.xbery.artbeams.jooq.schema.tables.references.ROLES
import org.xbery.artbeams.jooq.schema.tables.references.USER_ROLE
import org.xbery.artbeams.users.domain.Role
import org.xbery.artbeams.users.repository.mapper.RoleMapper
import org.xbery.artbeams.users.repository.mapper.RoleUnmapper

/**
 * Role repository.
 * @author Radek Beran
 */
@Repository
class RoleRepository(
    override val dsl: DSLContext,
    override val mapper: RoleMapper,
    override val unmapper: RoleUnmapper
) : AssetRepository<Role, RolesRecord>(
    dsl, mapper, unmapper
) {
    override val table: Table<RolesRecord> = ROLES
    override val idField: Field<String?> = ROLES.ID

    fun findRoles(): List<Role> =
        dsl.selectFrom(table).orderBy(ROLES.NAME).fetch(mapper)

    fun findRolesOfUser(userId: String): List<Role> =
        dsl.selectFrom(table)
            .where(ROLES.ID.`in`(
                dsl.select(USER_ROLE.ROLE_ID)
                    .from(USER_ROLE)
                    .where(USER_ROLE.USER_ID.eq(userId))
            ))
            .orderBy(ROLES.NAME)
            .fetch(mapper)

    fun updateRolesOfUser(userId: String, roles: List<Role>) {
        dsl.deleteFrom(USER_ROLE)
            .where(USER_ROLE.USER_ID.eq(userId))
            .execute()
        for (role in roles) {
            dsl.insertInto(USER_ROLE)
                .set(USER_ROLE.USER_ID, userId)
                .set(USER_ROLE.ROLE_ID, role.id)
                .execute()
        }
    }
}
